### ISSUE TITLE
Async daemon can lose events in a highly concurrent scenario

### ISSUE THREAD

**Issue by jacobpovar** [JasperFx/marten#1629]
This is detailed description of the Async daemon algorithm issue that I described to @oskardudycz  some time ago.  Sorry for the delay.

In short, when multiple threads are writing to the same event store, it's highly unlikely but possible for the daemon to loose some of these events.

## Problem

Async deamon, of specifically, fetcher, uses polling based on sequences of the events.

Fetcher [sends 4 quieries on each iteration](https://github.com/JasperFx/marten/blob/master/src/Marten/Events/Projections/Async/Fetcher.cs#L56):
- selects sequences of the next page starting from `_lastEncountered`, 
- then events itself within this page,
- and then the next sequence number and last known sequences in the event store.

Postgres sequence are unaware of ongoing transactions. So it's possible when multiple threads or processes are writing to the same database, that some of the events will be persisted out of order with respect to theirs sequence numbers.
For example `thread A` starts writing one huge event EventA with sequence == 100, and while it's writing `thread 2` writes tiny EventB with sequence == 101, and only after that EventA finishes writing.
As a result, EventB with larger sequence value gets written before EventA with small Sequence value.

Fetcher is _mostly_ aware of this behavior, so when reading the next page it checks page for gaps in it. If any - page [gets re-read](https://github.com/JasperFx/marten/blob/master/src/Marten/Events/Projections/Async/Fetcher.cs#L170) 

So far so good.

After the whole page has been read from the event store fetcher [reads sequence number of next event that has matching type](https://github.com/JasperFx/marten/blob/master/src/Marten/Events/Projections/Async/Fetcher.cs#L58). 

At the end, fetcher queue's page for processing and saves `NextKnownSequence - 1` as `_lastEncountered`. Code: 
[fetcher](https://github.com/JasperFx/marten/blob/0b5496ab860ffb857a766a5965a9f307b1fcc4fb/src/Marten/Events/Projections/Async/EventPage.cs#L64), [page](https://github.com/JasperFx/marten/blob/0b5496ab860ffb857a766a5965a9f307b1fcc4fb/src/Marten/Events/Projections/Async/EventPage.cs#L64)

The problem occurs when fetcher has read NextKnownSequence and another thread _after that_ finishes writing event with sequence number that is greater than last page End but smaller than NextKnownSequence. As a result this event gets skipped.
So there no gaps in event pages but gaps a possible **between** pages.

## Repro 
It can be reproduced by running multiple threads start daemon that saves processes sequences https://gist.github.com/jacobpovar/7c479de175a32afaa3d3b99d6f02f9cf. 
In my setup it doesn't occur every time, but when run multiple times misses will most likely happen.
This is the sample output from the repro run: 
```
12085 is missing.
12086 is missing.
12087 is missing.
12088 is missing.
12089 is missing.
12090 is missing.
12091 is missing.
12092 is missing.
12093 is missing.
12094 is missing.
12095 is missing.
12096 is missing.
12097 is missing.
12098 is missing.
12099 is missing.
12100 is missing.
12101 is missing.
43022 is missing.
43023 is missing.
43024 is missing.
43025 is missing.
...
...
**From: 12000, to: 12084, next 12102
From: 12101, to: 12201, next 12202**
From: 12201, to: 12301, next 12302
From: 12301, to: 12401, next 12402
From: 12401, to: 12501, next 12502
From: 12501, to: 12601, next 12602
....

```
and if we look into the database we can see that event 12102 has smaller timestamp and probably was written before preceding events.
![image](https://user-images.githubusercontent.com/664640/100624575-a9803600-3334-11eb-915f-c648133996bc.png)

## Possible solution 
A possible and simplest solution would by to start fetching next page just after the current, without querying for the NextKnownSequence. Daemon's performance would drop slightly however. And there is no much sense in doing that if daemon is reading historical events. Such thorough check is only worth it if daemon is up to date with the event store end. This is what we implemented at my work.

**Comment by jeremydmiller** [comment]
@jacobpovar I think this code is basically getting re-written almost from scratch in v4, but in your case, did you try to increase the "leading edge buffer"? That makes the async daemon purposely ignore events that are too new to allow the missing sequences to get filled in before it proceeds.

**Comment by jacobpovar** [comment]
@jeremydmiller haven't tried it but increasing leading edge buffer should help, thanks.

**Comment by jeremydmiller** [comment]
Shutting this down since it's rewritten in v4

**Comment by simbadltd** [comment]
Sorry for resurrecting the topic. But I've an idea for possible solution. Described it here: [Events loss in Marten.Events.Projections.Async.Fetcher under load #1879]().

**Comment by oskardudycz** [comment]
@jacobpovar could you check version 3.13.5? It should have the fix from #1879.

See more in: https://github.com/JasperFx/marten/releases/tag/3.13.5.

I'm closing the issue, if you're still facing the issue, I'll reopen it.
